clippy_lints/src/unbound_return_lifetimes.rs

```diff
@@ -0,0 +1,185 @@
+use if_chain::if_chain;
+use rustc::declare_lint_pass;
+use rustc::hir::map::Map;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc_hir::intravisit::*;
+use rustc_hir::*;
+use rustc_session::declare_tool_lint;
+
+use crate::utils::span_lint;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for lifetime annotations on return values
+    /// which might not always get bound.
+    ///
+    /// **Why is this bad?** If the function contains unsafe code which
+    /// transmutes to the lifetime, the resulting lifetime may live longer
+    /// than was intended.
+    ///
+    /// **Known problems*: None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// struct WrappedStr(str);
+    ///
+    /// // Bad: unbound return lifetime causing unsoundness (e.g. when x is String)
+    /// fn unbound<'a>(x: impl AsRef<str> + 'a) -> &'a WrappedStr {
+    ///    let s = x.as_ref();
+    ///    unsafe { &*(s as *const str as *const WrappedStr) }
+    /// }
+    ///
+    /// // Good: bound return lifetime is sound
+    /// fn bound<'a>(x: &'a str) -> &'a WrappedStr {
+    ///   unsafe { &*(x as *const str as *const WrappedStr) }
+    /// }
+    /// ```
+    pub UNBOUND_RETURN_LIFETIMES,
+    correctness,
+    "unbound lifetimes in function return values"
+}
+
+declare_lint_pass!(UnboundReturnLifetimes => [UNBOUND_RETURN_LIFETIMES]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnboundReturnLifetimes {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx Item<'tcx>) {
+        if let ItemKind::Fn(ref sig, ref generics, _id) = item.kind {
+            check_fn_inner(cx, &sig.decl, generics, None);
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'a, 'tcx>, item: &'tcx ImplItem<'tcx>) {
+        if let ImplItemKind::Method(ref sig, _) = item.kind {
+            let parent_generics = impl_generics(cx, item.hir_id);
+            check_fn_inner(cx, &sig.decl, &item.generics, parent_generics);
+        }
+    }
+}
+
+fn check_fn_inner<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    decl: &'tcx FnDecl<'tcx>,
+    generics: &'tcx Generics<'tcx>,
+    parent_data: Option<(&'tcx Generics<'tcx>, &'tcx Ty<'tcx>)>,
```

Comment:
Can inner functions refer to the lifetimes of outer functions? If yes, going one level up the stack might not be enough.